### ISSUE TITLE
Extension wishlist

### ISSUE THREAD

**Issue by minad** [minad/corfu#103]
- [x] corfu-history (Prototype in https://github.com/minad/corfu/tree/extensions)
- [x] corfu-info (Prototype in https://github.com/minad/corfu/tree/extensions, Move out `corfu-show-documentation/location` from corfu.el, make it persistent and live updating)
- [x] corfu-quick (#127)
- [x] corfu-indexed (#127)
- [x] corfu-doc by @galeo, see https://github.com/galeo/corfu-doc/issues/8. Merged as corfu-popupinfo.
- [ ] corfu-mouse (See https://codeberg.org/akib/emacs-corfu-mouse, waiting on @AkibAzmain's copyright assignment)
- [ ] corfu-narrow (Proposal in #147)

**Comment by Luis-Henriquez-Perez** [comment]
I take a similar approach to `vertico-quick` and modify the `corfu--format-candidates` function. I add `quick-letters` keys to the prefix of each candidate and am calling `corfu--format-candidates` with the updated candidates.

```elisp
(defun corfu-quick--format-candidates (orig candidates)
  "Format candidate, see `corfu--format-candidate' for arguments."
  (let ((updated-candidates nil))
    (pcase-dolist (`(,candidate ,prefix ,suffix) candidates)
      (setq quick-letters (corfu-quick--letters))
      (push (list candidate (concat quick-letters " " prefix) suffix) updated-candidates))
    (cl-callf reverse updated-candidates)
    (funcall orig updated-candidates)))
```

However, I'm having some trouble updating the `corfu` popup to display the candidates prefixed by the quick keys. What's `corfu`'s counterpart to `vertico--exhibit`? I thought it would be `corfu--update`. However, even though the candidates format is correctly updated by `corfu-quick--format-candidates` the popup disappears after calling `corfu-update`.

```elisp
(defun corfu-quick--read ()
  "Read quick key given FIRST pressed key."
  (cl-letf* ((old-fn (symbol-function #'corfu--format-candidates))
	     (new-fn (apply-partially #'corfu-quick--format-candidates old-fn))
	     ((symbol-function #'corfu--format-candidates) new-fn))
    (message "RESETTING")
    (corfu--update)
    1))
```

**Comment by minad** [comment]
> I take a similar approach to vertico-quick and modify the corfu--format-candidates function. I add quick-letters keys to the prefix of each candidate and am calling corfu--format-candidates with the updated candidates.

Sounds good. Please overwrite the prefix with the quick letters as Vertico does to save space. But we have to make sure that this works nicely with @jdtsmith's kind-icon.

>  thought it would be corfu--update. However, even though the candidates format is correctly updated by corfu-quick--format-candidates the popup disappears after calling corfu-update.

`corfu--update` is not an exact equivalent to `vertico--exhibit`, since it also contains some termination logic. I think you should call `corfu--candidates-popup` directly, which is called by `corfu--update`. We can also restructure this code in Corfu itself to make it easier to hook in the quick extension. But for now I'd prefer if we work with what we've got.

**Comment by minad** [comment]
@Luis-Henriquez-Perez One more thing - my recommendation would be to first implement `corfu-indexed`. I know it is not the extension you are interested in and I am neither. However if we provide `corfu-quick` we should also provide `corfu-indexed`, that's at least the plan. It should be much easier to write this extension and get it finished.

**Comment by Luis-Henriquez-Perez** [comment]
> Please overwrite the prefix with the quick letters as Vertico does to save space. But we have to make sure that this works nicely with @jdtsmith's kind-icon.

I don't mind overwriting it instead. The reason I appended to the prefix instead of overwritting it precisely because I was afraid another package could be using these prefixes such as `kind-icon`.

> my recommendation would be to first implement corfu-indexed.

>  I'd prefer if we work with what we've got.

I agree.

> my recommendation would be to first implement corfu-indexed. I know it is not the extension you are interested in and I am neither.

O.K. I'll switch my attention to `corfu-indexed`.

**Comment by jdtsmith** [comment]
How many characters are the quick keys?  Always one?  If not, you'd need to arrange for them to be the same width always to avoid formatting issues.  It might be nice to inherit the BG color from the existing prefix as well (if any).

**Comment by minad** [comment]
> How many characters are the quick keys? Always one? If not, you'd need to arrange for them to be the same width always to avoid formatting issues. It might be nice to inherit the BG color from the existing prefix as well (if any).

This formatting detail is not important right now. We can address this when we have something which works.

**Comment by jdtsmith** [comment]
Company kinds are quite finite and stable (as they are hard-coded into company).  But rather than exhaustive, a corfu-narrow extension could implement "just a few" narrowing targets based on the extension user's config.  I.e. "show me variables"; "show me functions"; "show me everything".  IMO, this is the kind of "only on corfu functionality" that could set it apart.  Would be good to simplify this messaging.  IMO, corfu uniqueness points are 1) works with CAPFs exclusively (but some may see that as a drawback), 2) works anywhere completion-at-point is invoked, 3) mini-frame based, 4) superior support for advanced completion styles like orderless.

I might be able to play with this; where would you suggest injecting/wrapping a kind-narrowing predicate into corfu?

**Comment by minad** [comment]
@jdtsmith I am not opposed to adding narrowing as an extension. However the extensions are currently on hold. It needs a bit of work to push this over the edge and my focus isn't on Corfu right now. There are a bunch of extensions which seem almost ready: corfu-history from the history branch, corfu-indexed and corfu-quick by @Luis-Henriquez-Perez. Also moving the documentation functions out of corfu.el to corfu-help.el is a quick thing. This work should come first.

Regarding the uniqueness points: 1) is factually wrong, since Corfu enjoys **full** Company backend support via `cape-company-to-capf`. This is not a drawback in comparison to Company since Company also needs the `company-capf` adapter in the other direction. The most important uniqueness point however is the simplicity and better integration. This is where my focus is. Therefore I don't want to add anything to the core implementation itself. All kinds of experiments and further enhancements should go into extensions.

**Comment by AkibAzmain** [comment]
How about Corfu in terminal?  I've rewritten `corfu--popup-show` and `corfu--popup-hide` to show popup on terminal.  It has an additional dependency however, [Popon](https://codeberg.org/akib/emacs-popon), which I plan to release on ELPA.  Would you like to accept [it](https://codeberg.org/akib/emacs-corfu-popup) as an extension?  Maintaining code copied from another package is a burden, especially when doing that just to remove `display-graphic-p` calls.

**Comment by minad** [comment]
@AkibAzmain Oh that's great! We cannot add it to Corfu if it has an additional dependency. But it is no problem to release this separately, similar to `vertico-posframe`. We can try to make sure that the APIs don't break. The base functionality of Corfu/Vertico is quite stable so breakage is unlikely. I recommend to modify your advices as follows:

~~~ elisp
(defun corfu-popup--in-region (&rest app)
  (cl-letf (((symbol-function #'display-graphic-p) (lambda () t)))
    (apply app)))

(advice-add #'corfu--in-region :around #'corfu-popup--in-region)
~~~

Maybe we can come up with a better integration strategy later on (which reduces the amount of advices), but for now I think this is good enough until your package stabilizes.

**Comment by AkibAzmain** [comment]
> @AkibAzmain Oh that's great! We cannot add it to Corfu if it has an additional dependency. But it is no problem to release this separately, similar to `vertico-posframe`.

What's the problem?  Legal issues?  If that's the case then releasing Popon to ELPA should solve it.  And if the problem is the dependency itself then I have no problem maintaining it separately, but I'm curious to know why you only depend on base Emacs.  Previously `corfu-popup` had its own popup implementation, but then I realized that it's impossible to show multiple popups at the same time (i.e showing documentation popup while Corfu popup is visible will break both), and the solution was to use a package dedicated to this task.  As I found no such package on ELPA (there's one on MELPA, but that's not copyrighted by FSF), I wrote a new one.

> We can try to make sure that the APIs don't break. The base functionality of Corfu/Vertico is quite stable so breakage is unlikely. I recommend to modify your advices as follows:
> 
> ```emacs-lisp
> (defun corfu-popup--override-display-graphic-p (&rest app)
>   (cl-letf (((symbol-function #'display-graphic-p) (lambda () t)))
>     (apply app)))
> 
> (advice-add #'corfu--in-region :around #'corfu-popup--override-display-graphic-p)
> (advice-add #'corfu--auto-post-command :around #'corfu-popup--override-display-graphic-p)
> ```
> 
> Maybe we can come up with a better integration strategy later on (which reduces the amount of advices), but for now I think this is good enough until your package stabilizes.

I considered that option, but didn't do that because some function may really need (e.g. `corfu-margin-formatters` maybe) it.

**Comment by minad** [comment]
> What's the problem?

There is no technical problem. I just don't want to have a dependency for Corfu. Since I can provide all the functionality of Corfu without dependencies, I avoid them. If I would really need a dependency I would add it. 

Anyway, to be clear about the terminal display - I am not the best person to maintain it for a few reasons and due to my preferences. I am a proponent of using GUI Emacs and I use GUI Emacs exclusively. As you probably saw, Corfu started as an overlay-based popup (Completion **Overlay** Region FUnction), but I've given up on that one due to the amounts of bugs caused by it. Then I decided to embrace child frames and take advantage of the additional styling possibilities. In the future, we could profit even more from child frames by supporting variable pitch fonts. This is also relevant for some languages where characters are often not of fixed pitch. Furthermore maintaining the child frame display alone is already painful. For all these reasons it will be much better if the terminal popup code is maintained separately from Corfu by someone who actually uses this part of the code. I simply won't be able to guarantee sufficient quality.

I really appreciate that you are working on Popon and a Corfu display using Popon. But to be honest - ultimately I think the popup issue on terminals should be solved on the level of the display engine instead of with overlays. This would then make the functionality significantly easier to maintain. Interestingly the terminal popup functionality is already there but just not exposed and made available properly. Try the following with emacs -Q -nw, which despite the name surprisingly works:

~~~ elisp
(x-popup-menu `((,(car (posn-x-y (posn-at-point (point))))
                 ,(+ 3 (cdr (posn-x-y (posn-at-point (point))))))
                ,(selected-frame))
              (easy-menu-create-menu
               "Menu"
               `(["item 1" item1]
                 ["item 2" item2]
                 ["item 3" item3])))
~~~

> I considered that option, but didn't do that because some function may really need (e.g. corfu-margin-formatters maybe) it.

I don't think this is a big deal given that you can disable the margin formatters, e.g., you could also let-bind them to nil. The `corfu-popup--override-display-graphic-p` is certainly better than copying the functions verbatim from Corfu.

**Comment by minad** [comment]
@AkibAzmain Did you consider naming your package `corfu-terminal` instead? I think then it would be much more obvious what it is about.

**Comment by AkibAzmain** [comment]
> Did you consider naming your package `corfu-terminal` instead?

No, I didn't consider that.  However, it works even on GUI, and looks mostly like the child frame UI.

Corfu (default) | `corfu-popup`
--|--
[![2022-05-06-212036-447x220-scrot.png](https://i.postimg.cc/MH6d3JWs/2022-05-06-212036-447x220-scrot.png)](https://postimg.cc/sQLY18GS) | [![2022-05-06-212133-454x226-scrot.png](https://i.postimg.cc/brVH6GVJ/2022-05-06-212133-454x226-scrot.png)](https://postimg.cc/30pvNwT5)

**Comment by jdtsmith** [comment]
This looks great!

**Comment by AkibAzmain** [comment]
@jdtsmith I have add some screenshots in [README](https://codeberg.org/akib/emacs-corfu-popup) along with installation instructions, check them out!

**Comment by jdtsmith** [comment]
Thanks.  For the GUI in kind-icon I use half-spaces on each side for text-based badges when they are a single letter.  Do you know how emacs on the terminal treats partial spaces like `(propertize " " 'display `(space :width (,(/ (default-font-width) 2))))`?  Never shows them, shows them as a full space if they are >=1/2 space width, etc? What does `(default-font-width)` eval to for you?  I think we could improve this handling; feel free to open an issue.

**Comment by AkibAzmain** [comment]
@jdtsmith,
> Thanks. For the GUI in kind-icon I use half-spaces on each side for text-based badges when they are a single letter.

> What does `(default-font-width)` eval to for you (1 I guess)?

`1`.

> Do you know how emacs on the terminal treats partial spaces like `` (propertize " " 'display `(space :width (,(/ (default-font-width) 2)))) ``? Never shows them, shows them as a full space if they are >=1/2 space width, etc?

Emacs shows any N pixels width space as N space characters on terminal (This is what I my Emacs does, don't know what the manual says). `(/ (default-font-width) 2)` = `(/ 1 2)` = `0` (integer, not float), so nothing is visible.

> I think we could improve this handling on terminal by right padding with full space for two char badges, and padding on both sides for 1; feel free to open an issue.

Yeah, some workarounds are needed to make `kind-icon` work on terminal, specially on tty.

**Comment by jdtsmith** [comment]
> some workarounds are needed to make kind-icon work on terminal, specially on tty

I pushed a commit that right-pads two character text icons on terminals (or anywhere with default-font-width = 1). Give it a try and let me know how that works?

**Comment by AkibAzmain** [comment]
@jdtsmith Ok, I'll try that.  Waiting for elpa-dev update.

**Comment by AkibAzmain** [comment]
@jdtsmith, Looks nice on terminal.  But on GUI multi-character text icons don't work well with overlays, as replacing spec is used as display property, which uses the face of the original text in the buffer.  I have a solution, replace the following in `kind-icon-formatted`:

```emacs-lisp
;; text, 1 or 2 chars, centered with full or half space on each side
(let* ((txt (truncate-string-to-width (cadr map) 2))
       (len (length txt))
       (txt-disp (if (eq len 2)
                     (if terminal
                         (concat txt " ")
                       (concat pad-left
                               (propertize "_" 'display txt)
                               pad-right))
                   (concat " " txt " "))))
  (propertize txt-disp 'face
              `(:weight bold :foreground ,col :background ,bg-col)))
```

With this:

```emacs-lisp
(let* ((txt (truncate-string-to-width (cadr map) 2))
       (len (length txt))
       (txt-disp (if (eq len 2)
                     (if terminal
                         (concat txt " ")
                       (concat pad-left
                               (propertize
                                "_" 'display
                                ;; Propertize the display text itself too
                                (propertize
                                 txt 'face
                                 `((:weight bold
                                            :foreground ,col
                                            :background
                                            ,bg-col)
                                   ;; Maybe put `kind-icon-default-face'
                                   ;; in fallback list too?
                                   default)))
                               pad-right))
                   (concat " " txt " "))))
  (propertize txt-disp 'face
              `(:weight bold :foreground ,col :background ,bg-col)))
```

With the above text icons on GUI using popons look as good as using child frames.

**Comment by jdtsmith** [comment]
Interesting. This sounds like an emacs bug to me, and I don't see that in emacs-28 terminal:

What does:

```elisp
(let ((txt-disp (concat " " (propertize "_" 'display "foobar") " ")))
  (insert "\n"
	  (propertize txt-disp 'face
		      `(:weight bold :foreground "red" :background "yellow"))))
```

show for you on terminal?  I always get the right colors.

**Comment by AkibAzmain** [comment]
@jdtsmith, Your snippet shows the right colors, *because you insert the text*.  But when you put the string in the `before-string` or `after-string` of an overlay, colors will be wrong on both GUI and terminal.  This may be a Emacs 27 bug (or feature?) which may not be present in Emacs 28.  Here is the example code which doesn't show the right colors:

```emacs-lisp
(let ((txt-disp (concat " " (propertize "_" 'display "foobar") " "))
      (ov (make-overlay (point-min) (point-max))))
  (overlay-put ov 'before-string
	       (propertize txt-disp 'face
		           `(:weight bold :foreground "red" :background "yellow"))))

;; To remove the overlay call `delete-all-overlays`.
```

And the workaround is to set the `face` property of the display text itself.

**Comment by jdtsmith** [comment]
I see, that indeed is not improved in Emacs 28. Let me look at it more closely. Can you open an issue on kind-icon with the details?

**Comment by AkibAzmain** [comment]
@jdtsmith, I have opened a new issue, jdtsmith/kind-icon#16.

**Comment by liuyinz** [comment]
@AkibAzmain , there is a bug when enable corfu-popup-mode just as the gif below:

![Kapture 2022-05-12 at 05 51 00](https://user-images.githubusercontent.com/35212403/167954014-a2ef7755-1dae-4cf2-a851-a36e91a38051.gif)
 
To reproduce bug : 
1. The texts which covered by popup menu, must have face with background color or inverse-video:t.
2. The texts must start at line beginning,such as position (xxx:0).

**Comment by minad** [comment]
@liuyinz Please report this on the corfu-popup tracker.

**Comment by AkibAzmain** [comment]
@liuyinz I can't reproduce this on Emacs 27.2.  Are there any overlays in the buffer?  Can you open an issue [here](https://codeberg.org/akib/emacs-corfu-popup/issues/new) with more details?

**Comment by liuyinz** [comment]
@AkibAzmain [here](https://codeberg.org/akib/emacs-corfu-popup/issues/2)

**Comment by minad** [comment]
What about the name corfu-popon as suggested on emacs-devel?

**Comment by minad** [comment]
@AkibAzmain I would be fine with both names but given that you also proposed popon, flymake-popon and the corfu-popup/terminal/popon package, using -popon is more consistent. Maybe you are going to add even more -popon packages, e.g., flyspell-popon, ispell-popon?

**Comment by AkibAzmain** [comment]
@minad,

> @AkibAzmain I would be fine with both names but given that you also proposed popon, flymake-popon and the corfu-popup/terminal/popon package for ELPA, using the -popon suffix seems more consistent.

I decided to name it `corfu-terminal` because it's purpose is to make Corfu usable on terminal (or any other non-graphical display), not `corfu-popon` because Popon is new word invented by me, which most users won't understand and most users usually want things just work and don't care about the implementation.  But if you think that **consistency** more important, please inform me.

> Maybe you are going to add even more -popon packages, e.g., flyspell-popon, ispell-popon?

Maybe, but then I will name them *-terminal if already a better GUI package exists and I don't change the name of `corfu-terminal` to `corfu-popon`, for consistency.

**Comment by minad** [comment]
@AkibAzmain Okay, got it. I am fine with the name corfu-terminal. I just extracted the function corfu-popup-support-p in https://github.com/minad/corfu/commit/cf58c19598aee9828da49bdfb7d596208a91ccc9. This will help you to avoid the function patching dance you have in corfu-terminal. Now you only have to advice corfu--popup-show/hide/support-p to replace the display code. Maybe you could even add a check to corfu--popup-show/hide to fall back to the original child frames if on graphical display and only use the terminal code if in terminal, such that the user always gets the best frontend? This fallback behavior could also be configurable with a variable corfu-terminal-prefer-child-frame.

**Comment by BooAA** [comment]
@minad I wish corfu can support interface similar to `ivy-completion-in-region` or vim-style completion. See the screencast below:

https://user-images.githubusercontent.com/32809182/169704794-1a8d3ef6-ae34-4a4d-ad1a-b1fe08a2fed3.mov

The key here is that after manually trigger completion, we go into a session that can edit the input freely (insert space, `M-DEL` to kill a word, or even remove the whole string and search for a new one). The completion session will terminate only when users explicitly end it themselves (in ivy's case, hitting `RET` will submit the candidates, otherwise the session keep going even nothing is matched or the search string is totally different).

This only makes sense when the user manually trigger the completion, but with this functionality, we no longer worried about quit the completion session unexpectedly when editing the input. For example, when using corfu with tab-and-go style, after using tab several times to insert the candidate, I cannot backspace to kill some characters and keep my completion session. 

https://user-images.githubusercontent.com/32809182/169705491-62a59995-e169-4538-8080-d726b23812c0.mov

In vim, they provide similar feature with `C-p` that support tab-and-go and also be able to edit the input when something is already insert or when input is mismatched, only `ESC` or `SPC` will quit the completion session.

https://user-images.githubusercontent.com/32809182/169705832-a64964c4-8d26-4830-b560-a5a817bdc9ea.mov

Notice that in the video, after some candidates are inserted, I can still backspace to edit the string. The session only ends when `SPC` or `ESC` are pressed. I'm not sure whether corfu can do this. In my experiment, I set `corfu-quit-no-match` and `corfu-quit-at-boundary` to `nil`, but still cannot reproduce behavior in vim or ivy.

**Comment by minad** [comment]
@BooAA If you use the following settings don't you get approximately the behavior that you are looking for? Can you be precise what is missing if you use these settings?

~~~ elisp
(setq corfu-quit-no-match nil
      corfu-quit-at-boundary nil
      corfu-preview-current nil)
~~~

**Comment by BooAA** [comment]
@minad Thanks, it's very close. Maybe I can tried to adapt it in my workflow. But I still want the tab-and-go behavior which saves lots of typing for me. Since I like using tab to insert some candidates and edit it for my need.

**Comment by jdtsmith** [comment]
@BooAA you may also need to use `cape-capf-buster` to wrap your CAPF if it does not correctly dump its cache when the input prefix changes (e.g. lsp's CAPF).

**Comment by BooAA** [comment]
@minad I tried to summarize the behavior in vim, which is a little bit complicated and I'm not sure whether there is an easy way to tune confu like that.

After a completion session starts:
1. If no candidate is inserted (e.g. no tab is pressed), it's ok to mismatch and the session won't exit.
2. After candidate is insert: 
a) if user remove some characters and keep edit the input, the session won't quit (even there is no match).
b) If user directly append new characters right after insertion, then session exit.
3. If `SPC` or some other special characters are inserted, quit the session.

Hope this helps.

**Comment by minad** [comment]
@BooAA Okay, I understand. I think we cannot easily support the insertion with subsequent editing, while keeping the session alive. When tabbing (Tab-n-Go) the candidate is not actually inserted, but only previewed (corfu-preview-current). As soon as you make further input or press backspace, the previewed candidate is finally inserted, completion is terminated and then the further input is appended.

**Comment by AkibAzmain** [comment]
Hello, everyone.  I have written another new package: [`corfu-mouse`](https://codeberg.org/akib/emacs-corfu-mouse).  Strictly speaking, I didn't write it from scratch.  It is based on `vertico-mouse`.  How about adding this as an extension?

**Comment by minad** [comment]
@AkibAzmain I'd be happy to accept this if it would work. Unfortunately this is not as trivial as the vertico-mouse package, otherwise I would have added this already. Also I am worried about the ignore map. I only introduced this map as a hack to prevents the child frame from taking focus.

**Comment by AkibAzmain** [comment]
> Edit: Sorry, although scrolling works, but mouse click somehow doesn't work (the completion session terminates just before handling the mouse click).

Sorry for spreading another misinformation.  Scrolling also didn't work when I wrote that.  But the good news is now both scrolling and mouse click work.  However, mouse click doesn't work when the preselected candidate is not selected.  @minad, Can you help me about this?

> I'd be happy to accept this if it would work. Unfortunately this is not as trivial as the vertico-mouse package, otherwise I would have added this already.

You are right @minad, because it involves handling events in multiple frames.

> Also I am worried about the ignore map. I only introduced this map as a hack to prevents the child frame from taking focus.

I have workarounded that mouse ignore map by binding the scroll events to nil, so that `mwheel-scroll` is called on those events.

**Comment by minad** [comment]
@AkibAzmain No, I would have to start again at the same point as you.

**Comment by AkibAzmain** [comment]
> However, mouse click doesn't work when the preselected candidate is not selected.

I have fix this problem.

The reason of the problem was that `corfu--pre-command` was terminating the completion session, because of the following check: `(/= corfu--index corfu--preselect)`.  @minad, thanks for *not* helping me, I just learned how to variable watchers can be useful for debugging.

**Comment by AkibAzmain** [comment]
> @AkibAzmain If you feel it is ready, please open a PR.

Yeah, it looks like it's ready.  But the only problem is paperwork, which I haven't done yet.

**Comment by AkibAzmain** [comment]
@BooAA Maybe a function like the following might help:

```emacs-lisp
(defun my-corfu-insert-for-tab ()
  "Insert current candidate, but don't end session."
  (interactive)
  (corfu--insert nil))
```

Note that I haven't tested this, but this should work according `corfu--insert` docstring and definition.

**Comment by AkibAzmain** [comment]
@milanglacier `corfu-terminal` and `corfu-doc-terminal` has an additional dependency named Popon (used to show the popup), which is a blocker for adding these to Corfu.

**Comment by minad** [comment]
@milanglacier @AkibAzmain I think the ultimate goal is to have child frame (or something child frame like) natively supported by Emacs on terminal. Then Corfu can support the terminal display more trivially. Until that happens, corfu-terminal will remain an external package.

**Comment by AkibAzmain** [comment]
@minad, is there any way to add `corfu-mouse` to Corfu without GitHub PR?  I don't want feed my code to Microsoft's code thief bot.

**Comment by minad** [comment]
@AkibAzmain You will feed it anyway since the code will be duplicated here. I understand your sentiment but I find these ideological discussions tiring. Please open a PR.

**Comment by minad** [comment]
@AkibAzmain I can only add corfu-mouse here if the copyright is assigned to the FSF.

**Comment by AkibAzmain** [comment]
@minad, I know that.  But what to do for now?  Or should I open the PR after assigning copyright?

**Comment by minad** [comment]
@AkibAzmain Yes, please open the PR after the copyright assignment has been completed.

**Comment by AkibAzmain** [comment]
> corfu-doc by @galeo, see Release on ELPA/MELPA? galeo/corfu-doc#8. Merged as corfu-infoframe.

But the name has change (again!).  :laughing:

**Comment by jdtsmith** [comment]
## Wish

A very simple and limited form of async support facilitating _cache pre-warming_ of completions.  

## Current situation

Right now corfu-auto sets up a timer, and, when the timer ticks over, consults the CAPFs.  If a CAPF says it can complete there, corfu requests the completions, and then pops them up.  But there is a problem when the CAPF table is gathering completions from an external process — it usually doesn't yet _know_ the completions.  But corfu urgently needs them.  The only choice the CAPF's table function has is to BLOCK until the process gives it what it needs.  There is not option to tell corfu: "I'll get back to you".

For many situations the block is unnoticeable (ms or 10s of ms), but sometimes (depending on process vagaries), it could be much more (2s sometimes in my usage).  To help, CAPF authors can do 2 things:

1. Make this blocking process call interruptible (inhibit-quit=nil),
2. Use `while-no-input` to abort on input.  

At least in this way the user is not frozen indefinitely.

## Problem

`while-no-input` has at fastest a ~1sec response on my system.   And after you type another character in... well the process starts all over again.  Typing with 1sec between keystrokes feels awful.  So high latency CAPFs get tossed overboard.

## Simple idea

A mode can provide to corfu an optional function via a special CAPF return _property_: `:before-complete` [1].  

If such a function is configured in the returned CAPF results, the process goes like:

1. corfu's timer function calls the `:before-complete` function with a callback function (which can capture the tick data via a closure var). 
5. On the CAPF side, after it gets the completion data it needs, the CAPF calls the callback function corfu supplied it, with `t` for success or `nil` for failure (optional: it can also just not call-back for failure).
6. In the callback itself, if the `:before-complete` function succeeded (and the tick data match), it goes on to call `corfu--auto-complete-deferred`, popping up completions, just as normal (if a bit delayed).
7. This _still_ checks the buffer tick to make sure nothing has changed, just like before.  

An analogous route for "on-demand" completion (`M-Tab`) completion would also be straightforward (though perhaps not necessary and unintuitive if it never pops up).

## Why?

With this simple support, a CAPF that routinely has high latency can configure a `:before-complete` function which pre-warms its cache and then calls back to corfu to continue, once it has (presumably asynchronously) gathered the relevant information. 

### Outcomes 

**What this does not do**: it does not decrease the actual latency before completions pop-up, for high latency processes. 

**What this does do**: it permits CAPF designs which do not block user input at all, as completion data are awaited fully asynchronously, if the CAPF is designed with latency in mind.

**What about other UIs**: They will still work.  They will ignore `:before-complete` and just directly call the CAPF table function, which will block (and cause people to say: just use corfu).

**Effects users may notice**: It may sometimes take longer for corfu to popup than usual.

**Complexities**: Suppose the timer ticks over and a CAPF's `:before-complete` function has not yet called the callback.  If buffer-tick is unchanged from the one sent to `:before-complete` there is nothing to do, just keep waiting. If the tick _has_ changed, well the old attempt can safely be ignore, and a new completion started.  A good CAPF will notice a new request to `:before-complete` coming in and cancel any that are pending, but that's not essential.

[1] could prefix it like `:corfu-before-complete` if you wanted to parallel company.

**Comment by minad** [comment]
I close this since there is not much movement here. Additional feature requests can go into separate issues.
